Homie 3.0.0 & 4dev convention controller Plugin

Christian-Me opened the pull request:
This is the 1st Version of a Homie convention controller plugin.
for more information [Homie - An MQTT Convention for IoT/M2M](https://homieiot.github.io/)

- sending "complete" auto-discover information (in openhab no typing required) not working reliable with build in broker due to retain=1 messages via QoS=0 issue (internal broker ignores the retain flag if a messages arrivis with QoS=0). With an external broker like mosquitto on default settings it should work. 
- There are known issues with MQTT Homie binding in 2.4 stable. After restart (or a longer period of time) openHAB get silent (not sending or receiving updates). 2.5 snapshot builds should resolve that issue. [see on openHAB forum](https://community.openhab.org/search?q=homie)
- can be compiled as homie 3.0.0 or 4.0.0 default to 4.0.0 because 4.0.0 do not require as many topics as 3.0.0
- providing a `homie/%sysname%/SYSTEM/cmd/set` topic for sending commands
- providing a homie/%sysname%/SYSTEM/gpio#/set  , topic to handle gpio switches (if port is set to default LOW or HIGH)
- dummy devices can receive value updates (new command DummyValueSet with same Parameters as TaskValueSet now build into P033_Dummy.
- most (only tested a few) plugin and event commands can be send over .../SYSTEM/cmd/set topic
- all devices which worked with the openHAB plugin should be able to send updates
- changing the unit name will send "disconnect" message and reconnects to the new topics and send new auto-discover messages. New unit (thing) should pop up in the inbox instantly. (old thing will not be deleted)

some known issues / ToDos

- [X] Acknowledgments should be integrated as controller command
- [X] good hook for "CPLUGIN_GOT_CONNECTED" function to send once after reboot
- [X] good hook for "CPLUGIN_FLUSH" before sleep mode
- [X] good hook for "CPLUGIN_GOT_DISCONNECT" if a device is renamed to delete the old
- [X] good hook for "CPLUGIN_INTERVAL" (if necessary because only for homie 3.0)
- [X] good hook for "CPLUGIN_ACKNOWLEDGE" to be triggered for all state changes to keep the controller up to date.
- [ ] autodiscover should provide units (needs a global concept for handling Units in ESPEasy)
- [ ] more actors to be implemented if useful
- [ ] update auto-discover data if devices are added or changed (currently you have to perform a reset)
- [ ] more ... 

![image](https://user-images.githubusercontent.com/15779507/55594690-d8e63600-5740-11e9-996b-76bcd628378c.png)

![image](https://user-images.githubusercontent.com/15779507/55594801-4db97000-5741-11e9-9d86-b765f2ff8fd0.png)

![image](https://user-images.githubusercontent.com/15779507/55594758-2662a300-5741-11e9-82d5-96560bdc599e.png)

My first pull request ;)

Christian-Me committed:
For 1st pull request

Should work with "complete" autodiscover, dummy device receive and many commands over SYSTEM/command/set topic

TD-er commented on src/_P033_Dummy.ino:
```diff
@@ -98,6 +98,37 @@ boolean Plugin_033(byte function, struct EventStruct *event, String& string)
+        String command = parseString(string, 1);
+        if (command == F("dummyvalueset"))
+        {
+          int deviceNr=parseString(string, 2).toInt(); // Index from 1-12;
```
The parameters of a command should already be available in the event.
If not, then they definitely should be parsed there and not in each and every plugin that's handling commands.
For example what happens if this is called with a parameter which cannot be converted into an integer?

Christian-Me replied:
You are right. As I was not sure if writing a PLUGIN_WRITE function for the dummy plugin I did a quick prototype. I'm not familiar with the event structure so I parsed everything out of the command. Now I'm using event->Par1(-5) to get the parameters.

TD-er commented on src/_P033_Dummy.ino:
```diff
@@ -98,6 +98,37 @@ boolean Plugin_033(byte function, struct EventStruct *event, String& string)
+            String valueStr=parseString(string, 4);
+            if (valueStr=="true") value=1;
+              else if(valueStr=="false") value=0;
+              else value=valueStr.toFloat();
```
Please use { } to mark the scope of an if-statement.
Especially when there is an else scope and even more in this case with if...else if...else.

Also use `F()` macro for strings, or else they will be placed in RAM.

And there is also a function to check if it is a valid float.

Christian-Me replied:
DummyValueSet will never receive true of false so I erased that part.
Hope I found all missing {} and added all F() macros.
Use string2float() instead .toFloat

TD-er commented on src/_P033_Dummy.ino:
```diff
@@ -98,6 +98,37 @@ boolean Plugin_033(byte function, struct EventStruct *event, String& string)
+
+            UserVar[event->BaseVarIndex+valueNr-1]=value;
+
+            addLog(LOG_LEVEL_INFO,log);
```
Please wrap the creation of the log string into a check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/__CPlugin.ino:
```diff
@@ -18,6 +18,15 @@ static const char ADDCPLUGIN_ERROR[] PROGMEM = "System: Error - To much C-Plugin
+
+void CPluginDisonnect(void)
+{
+    CPluginCall(CPLUGIN_DICONNECT, 0);
```
Typo in `CPLUGIN_DICONNECT`

Christian-Me replied:
Corrected. Still need to find a good place to call this function. I think it is only suitable when the "unit name" is changed by the user to inform the controller that the old name is not valid anymore.

TD-er replied:
I think if any of the parameters you want to send in the discovery message is changed, you must send the message again.
And if it is about re-informing the changes, I think you should name it like that.

TD-er commented on src/__Plugin.ino:
```diff
@@ -1181,10 +1182,15 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
                 if (retval) {
+#ifdef USES_C014  // Homie 3 & 4dev Controller
+                  // check if ther is usefull information which should be sent to a controller too.
+                  // for TESTING! Not shure if here is the right place. Should perhaps be passed to all controllers in the future.
```
Please try to minimize spelling errors in comments too

- ther => there
- shure => sure

Christian-Me replied:
try my best ;)

TD-er commented on src/__Plugin.ino:
```diff
@@ -1196,6 +1202,10 @@ byte PluginCall(byte Function, struct EventStruct *event, String& str)
               delay(0); // SMY: call delay(0) unconditionally
+#ifdef USES_C014
+              // and here too?
+              CPlugin_014_Feedback(Function, &TempEvent, str);
```
What is this comment about? What does it refer to?

Christian-Me replied:
I was asking myself why `Plugin_ptr[x](Function, event, str)` is called twice.
As CPluginCall(CPLUGIN_ACKNOWLEDGE, event, str); is now implemented I removed this.

TD-er commented on src/_P033_Dummy.ino:
```diff
@@ -98,6 +98,37 @@ boolean Plugin_033(byte function, struct EventStruct *event, String& string)
+            UserVar[event->BaseVarIndex+valueNr-1]=value;
+
+            addLog(LOG_LEVEL_INFO,log);
+            success = true;
```
This success assignment should only be done if there was a successful conversion of the value.
So this part should also be in the if...else if ... else scope.

Christian-Me replied:
Done.

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+          log += tmppubname;
+          log += ' ';
+          log += value;
+          addLog(LOG_LEVEL_DEBUG, log);
```
Please wrap this log string creation in a check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+              addLog(LOG_LEVEL_INFO, log);
+            }
+
+          } else TempEvent.Source = 0; // something whent wrong
```
What's being done with the `TempEvent` after this?
Is the else needed?
I would suspect something like `success = false;`, but in the rest of this case success is not handled at all.

Christian-Me replied:
At first I tired to implement the acknowledgement via the TempEvent .Source as the JSON acknowledgement is send for TempEvent .Source=VALUE_SOURCE_MQTT. Think this was not the right way . Removed.

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+                cmd += event->String2; // line,pos,Text or command as payload
+                validTopic = true;
+              } else if (pluginID==33) // Plugin 33 Dummy Device
+              { // DummyValueSet,<task/device nr>,<value nr>,<value/formula (!ToDo) >, works only with new version of P033!
```
I am not a fan of handling plugin commands in a controller.
Why have plugin 12 and 33 to be handled here in the controller and not in other controllers?
Or is there another way to reroute any command from a controller to the normal way of handling commands?
Then we can make a single list of commands that can be handled originating from "outside" the ESPeasy scope as some kind of 'firewall'.
This way we have to look at every controller that handles specific plugin commands.

Christian-Me replied:
I agree here, my initial concept was and still is to use the homie protocol and it´s autodetect features to make integrating a diy device running ESPEasy as flawless as possible for an end user and experienced user too. If you defined the unit name, all your devices, all their values in ESPEasy it is quite convenient not doing the same in your controller software again trying to match all topics, value names and types ... 
As you wrote before ESPEasy started as a "sensor" firmware sending data to `.../%value`. But when it comes to actors like gpio, rgb leds like neopixel ... there must be a defined route to pass information back. Homie uses a simple` .../%value%/set` path to do that.
Later I found out that the dummy device is a good way to pass numeric data to ESPEasy. This then can be handled by rules and passed to devices as nessesary.
For all commands (or better commands protected by ESPEasy) there is the cmd path as used before (i.e. in the openHAB plugin) only now coming from homie/%unitname%/SYSTEM/cmd/set instead /%unitname%/cmd.

But for some commands (but not many) I think it can be useful to have a direct way from a controller item to a value within ESPEasy. I.e. a switch item to a gpio pin performing ON/OFF action. Therefor the controller plugin not really handling the command it is more **translating** it from its topic and payload to the corresponding command.
The dummy device can be nicely connected to a slider / dimmer to control the pwm duty cycle for a LED or setting other values.
The LCD was only a test device for strings I had sitting next to me, but I will drop this because it makes no sense. It need 3 values (line, position and text), and ON OFF or CLEAR are not realy useful to bbe integrated in an UI without code.

The criteria for this translation should be:

- usefull for user interaction with standard UI elements (boolean, numeric value, RGB(W), HSV, text(?), enum (perhaps for IR controls))
- (none or) only one parameter necessary.
- no security risk.

So there is not much left beside gpio, dummy or neopixel / RGB(W) dimmers to pass data. In the future there could be a way for plugins to advertise their receive value capabilities beside commands. If this is added they should provide the information for the `$format` parameter too:

```
$format | Device → Controller Describes what are valid values for this property:
from:to Describes a range of values e.g. 10:15. Valid for datatypes integer, float (good for sliders and gauges)
value,value,value for enumerating all valid values.  e.g. A,B,C or ON,OFF,PAUSE. Valid for datatypes enum (remote controls)
rgb to provide colors in RGB format e.g. 255,255,0 for yellow. (sorry no RGBW here)
hsv to provide colors in HSV format e.g. 60,100,100 for yellow. Valid for datatype color 
```

As the controller plugin still use the PluginCall() path as the original openHAB plugin does so I see no actual security risk. 

So I`ll  drop the LCD part and keep 

- homie/%unitname%/SYSTEM/cmd/set to pass commands
- homie/%unitname%/SYSTEM/gpio#/set for switches
- homie/%unitname%/%dummyDeviceName%/%valueName%/set for dummy devices

I`ll do some experiments with RGB neopixels and perhaps come up with a solution for port expanders like PCA9685 or PCF8574 (which I use a lot because of the limited GPIOs available). For PCF8574 I do not have a idea how to detect if one port expander is present and how the ports will be used. Doing a i2c scan and rely on the i2c address found is not the right way I think.

Sorry for the long comment but I think the overall concept of using the homie convention should be easy to use user experience  beside standardization. I know that everything is currently already working in some way or another (using rules, mqtt import plugin, rules inside the controler, matching naming and so on) but not in a convenient or consistent way.

TD-er replied:
For the port expanders there is some activity in the background.
In short, I want a general GPIO handler which can be extended to port expanders.
This will then allow other plugins to use these GPIO's too.
So please have a quick search first or open a new issue about that one.

And about the communication in "the other direction". Those will result in a rather big change.
For the inter-ESPeasy communication we have the p2p protocol (part integrated in C013) which I plan to extend further. This can be a start to see what is really needed for the 2-direction communication. I still feel like it should be a task of the controller, but on the other hand we do not want to do all work again for a new controller. So I'm not really sure about that one yet.

Christian-Me replied:
I understand. I changed the code that currently ONLY gpio and DummyVaueSet are handled via /set and acknowledgements. This works fine, even when the command comes from another source like rule or HTTP. 
But if a different command like LongPulse changes the gpio state there will be no acknowledgement and the controller will be out of sync. 
Therefore the acknowledgement call should be placed directly where gpios are addressed. If there is already some work in progress for 2 way communication we should be happy for now and handle these cases with rules:

```
On gpioSeconds do
  gpio,%eventvalue1%,1
  let,1,%eventvalue1%
  timerSet,1,%eventvalue2%
endon

On Rules#Timer=1 do  //When Timer1 expires, do
  gpio,[VAR#1],0
endon
```

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+              {
+                cmd = "SYSTEM not found";
+              }
+            } else if (nodeName == F(CPLUGIN_014_CMD_VALUE)) // legethy homie/system/cmd topic
```
This kind of nested if... else scopes is very hard to read.
They do have the proper indentation, but still it would make sense to have a single level of if...else checks for the `nodeName` and in there call a separate function to handle those cases.

Also, what is "legethy" in the comment? "legacy" ?

Christian-Me replied:
You are right. Thats when the code is growing ... Have to sort it. Was started with a quick modification of the original openHAB controller ended much bigger than expected. That's fine I'll clean it up.
As I used the original openHAB controller as a template and I was keeping as much code as possible. The "legacy" (code & comment) could and should be erased now.

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+          TempEvent.Source = VALUE_SOURCE_MQTT; // to trigger the correct feedback
+          bool validTopic = false;
+          int lastindex = event->String1.lastIndexOf('/');
+          const String lastPartTopic = event->String1.substring(lastindex + 1);
```
This `lastPartTopic` is created, but does not seem to be used?

Christian-Me replied:
same "legacy" as before.

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+        log+=F(" messages)");
+        msgCounter = 0;
+        errorCounter = 0;
+        addLog(LOG_LEVEL_DEBUG, log);
```
Try to wrap the log string creation in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+              } else { // device not enabeled
+                log = F("C014 : Device Disabled: ");
+                log += getPluginNameFromDeviceIndex(Settings.TaskDeviceNumber[x]);
+                addLog(LOG_LEVEL_INFO, log+" not propagated!")
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+                  } else { // Device has custom Values
+                    log = F("C014 : Device has custom values: ");
+                    log += getPluginNameFromDeviceIndex(Settings.TaskDeviceNumber[x]);
+                    addLog(LOG_LEVEL_INFO, log+" ?!")
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+        log += F(" T: ");
+        log += topic;
+        log += F(" P: ");
+        log += payload;
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+        String log = F("C014 : T:");
+        log += topic;
+        log += F(" P: ");
+        log += payload;
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+        log+=F(" cmd: ");
+        log+=str;
+        log+=F(" not implemented!");
+        addLog(LOG_LEVEL_ERROR, log);
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
done

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+      log += valueName;
+      log += F(" topic: ");
+      log += topic;
+      log += F(" value: ");
```
Wrap in check for `logLevelActiveFor`

Christian-Me replied:
Didn't notice that this exists, that's perfect, will do that for all logs. Will take a wile (there are many of them (perhaps too many).

Newbie question: What is the best way to do this? Do the changes locally and then commit?

TD-er replied:
Yep, you should continue on the branch you made on your local computer.
When you make a commit and push it, it will be appended to this pull request.

A pull request is about merging a branch into another branch.
So adding commits to the branch you want to merge will be included in this PR. (as soon as you push a commit)

Christian-Me replied:
So I hope I addressed all of your comments. I think it is time to commit the changes.
Let`s give it a try ;)

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+      log += valueBool;
+      log += F(" (");
+      log += valueInt;
+      log += F(")");
```
Wrap in check for `logLevelActiveFor`

TD-er commented on src/_C014.ino:
```diff
@@ -0,0 +1,686 @@
+      String topic = ControllerSettings.Publish;
+      topic.replace(F("%sysname%"), Settings.Name);
+      topic.replace(F("%tskname%"), CPLUGIN_014_SYSTEM_DEVICE);
+      topic.replace(F("%valname%"), CPLUGIN_014_GPIO_VALUE + toString(port,0));
```
Some of these template strings are also used in other string parse/replace functions.
Please check to see if they can be used or maybe conflict here.

Christian-Me replied:
Changed the template to the CPLUGIN_014_PUBLISH constant. To stay in the homie standard the user should not be able to change this. So there will be no conflict, I think.

TD-er commented on src/ESPEasy-Globals.h:
```diff
@@ -306,6 +306,13 @@ void check_size() {
+// new messages for autodiscover controller plugins (experimental) i.e. C014
+#define CPLUGIN_CONNECTED                  52 // call after connected to mqtt server to publich device autodicover features
+#define CPLUGIN_DICONNECT                  53 // should be called before major changes i.e. changing the device name to clean up data on the controller. !ToDo
+#define CPLUGIN_FALLING_ASLEEP             54 // call before falling asleep !ToDo
```
I have recently added a similar function to perform a flush.
Is that the same behavior?

TD-er replied:
Hmm, it has not yet been merged, but I meant this PR: https://github.com/letscontrolit/ESPEasy/pull/2426/files

Christian-Me replied:
Let`s start with an easy one (hope I do this right, that's all new for me, please be patient)

I think CPLUGIN_FALLING_ASLEEP will do the same as to your CPLUGIN_FLUSH  they could be equal or I use CPLUGIN_FLUSH and delete CPLUGIN_FALLING_ASLEEP.

```suggestion
#define CPLUGIN_CONNECTED                  54 // call after connected to mqtt server to publich device autodicover features
#define CPLUGIN_FALLING_ASLEEP             52 // call before falling asleep
```

TD-er replied:
I think CPLUGIN_FLUSH is more descriptive of what it does.
CPLUGIN_FALLING_ASLEEP is more about when it should be executed.

CPLUGIN_CONNECTED  could maybe also be renamed to `CPLUGIN_ON_CONNECT` or `CPLUGIN_GOT_CONNECTED`, which is more in line with the events used for WiFi connection states.

Christian-Me replied:
OK, fine will use `CPLUGIN_FLUSH` and `CPLUGIN_GOT_CONNECTED` (or is there already a event after MQTT connected I can use?)

perhaps like this?
```
#define CPLUGIN_FLUSH                      52
// new messages for autodiscover controller plugins (experimental) i.e. C014
#define CPLUGIN_GOT_CONNECTED              53 // call after connected to mqtt server to publich device autodicover features
#define CPLUGIN_GOT_INVALID                54 // should be called before major changes i.e. changing the device name to clean up data on the controller. !ToDo
#define CPLUGIN_INTERVAL                   55 // call every interval loop
#define CPLUGIN_FEEDBACK                   56 // call for sending feedback !ToDo done by direct function call in PluginCall() for now.
```

TD-er replied:
Not sure yet.
The other controllers do not have a lot of bi-directional communication.
So I don't think there is any other function that can be re-used for this.

Christian-Me replied:
OK!?

so how to do the following tasks better?
**CONNECTED:** Homie (and perhaps other conventions) need to send information about the properties of the unit after reboot.
**INVALID:** to inform that the controller can't expect further data under the this topic homie/%unitName%/#. i.e. when the unit name changes.
**INTERVAL:** to send a heartbeat. Currently ESPEasy works by a delay sending data every X seconds. That is normaly heartbeat enough. Homie 4 (development) do not need interval Information anymore. Homie 3 wants signal strength and uptime.

**FEEDBACK:** It is essential (for Homie) that after message to .../%valueName%/set a feedback to .../%valueName% is sent. A switch in openHAB stays in an intermediate state until it receives the feedback. Additionally i.e. GPIO changes (by a rule or other command source) should be sent to the controller to keep the values up to date). Currently a json formatted feedback is sent after a every mqtt comand setting a GPIO value which could/should be integrated into the corresponding controller.

Initially I thought that all functions which are specific to a controller should be handled by the controller plugin via these function IDs. I thought they do not harm any others. In the case of the json message implemented outside the plugin I can imagine controllers triggering errors (and openHAB is quite fast with "communication errors") when receiving incorrect messages.

TD-er replied:
Sounds good
Only the "feedback" I'm not sure about.
What about "acknowledge" ?

Christian-Me replied:
changed to CPLUGIN_ACKNOWLEDGE.

TD-er commented on src/Controller.ino:
```diff
@@ -245,10 +245,14 @@ bool MQTTCheck(int controller_idx)
         connectionFailures += 2;
       }
-      return MQTTConnect(controller_idx);
+      return (!MQTTclient.connected() || MQTTConnect(controller_idx)); // either connected above or now!?
```
Not sure about performing a connect here.
Either way your change does different things here when performing a connect. 
In the first call to connect it does call the `CPLUGIN_CONNECTED` function and in the other call it doesn't.

Other MQTT controllers are reconnected just fine, so this change may change current behavior for other controllers too.

Christian-Me replied:
I was not sure how `MQTTConnect(controller_idx);` works and when `MQTTclient_should_reconnect` is true. I thought that when `MQTTclient_should_reconnect `is true the function `MQTTConnect(controller_idx);` will be called twice.

Perhaps you know a better place for` CPluginCall(CPLUGIN_GOT_CONNECTED, 0);`
It should be triggered then a mqtt connection is established and stable (to send many) messages.

TD-er replied:
I have to look.
I think just setting `MQTTclient_should_reconnect` to true should initiate a new (re)connect.

Christian-Me replied:
```
      if (MQTTclient_should_reconnect) {
        addLog(LOG_LEVEL_ERROR, F("MQTT : Intentional reconnect"));
        // Connect first and call all installed controller to publish auto-discover data
        if (MQTTConnect(controller_idx)) CPluginCall(CPLUGIN_GOT_CONNECTED, 0);
      } else {
        connectionFailures += 2;
      }
      return MQTTConnect(controller_idx);
```

As I understood it on` MQTTclient_should_reconnect` it tries the first time to connect and on success it calls `CPluginCall(CPLUGIN_GOT_CONNECTED, 0);`
But on failure it will try again` return MQTTConnect(controller_idx);` but this time it will NOT call: `CPluginCall(CPLUGIN_GOT_CONNECTED, 0);` on success.

For me it`s fine. In the rare occasion this happens a manual reset helps and most of the time sending auto-discover is not necessary because the messages are retained in the (some:) broker. Only for new devices and modifications.

TD-er reviewed (changes requested):
I just too a quick look at the code and these are the first things that came to mind.

Christian-Me committed:
Updates on pull request

Adressed "most of" the comments.

Christian-Me committed:
Merge branch 'mega' into mega

TD-er commented on src/Controller.ino:
```diff
@@ -245,6 +245,10 @@ bool MQTTCheck(int controller_idx)
+        // Homie 3 & 4dev Controller Plugin
+        // Connect first and call all installed controller to publish autodiscover data
+        // hope this is the right place ;)
+        if (MQTTConnect(controller_idx)) CPluginCall(CPLUGIN_GOT_CONNECTED, 0);
```
You're still changing the existing connect routines for the MQTT controllers.
Outside this if scope is also a call to MQTTConnect. (`return MQTTConnect` ) Why not change that part to not immediately return the connect state but call the `CPluginCall(CPLUGIN_GOT_CONNECTED)` from inside that `MQTTConnect` function.
Then you know for sure it will always be called if a connect has been made.

Christian-Me replied:
I think I misunderstood you, sorry. Moved it to here. Think it is a "safe" place right after LWT is published.

```
  if (MQTTclient.publish(LWTTopic.c_str(), LWTMessageConnect.c_str(), 1)) {
    updateMQTTclient_connected();
    statusLED(true);
    mqtt_reconnect_count = 0;
    // call all installed controller to publish autodiscover data
    CPluginCall(CPLUGIN_GOT_CONNECTED, 0);
    return true; // end loop if succesfull
  }
```
 
Should I commit the change or wait for further comments?

Christian-Me replied:
I forgot your comment before, now I understood:

```
  if (MQTTclient.publish(LWTTopic.c_str(), LWTMessageConnect.c_str(), 1)) {
    updateMQTTclient_connected();
    statusLED(true);
    mqtt_reconnect_count = 0;
    // call all installed controller to publish autodiscover data
    if (MQTTclient_should_reconnect) CPluginCall(CPLUGIN_GOT_CONNECTED, 0);
    return true; // end loop if succesfull
  }
```

TD-er replied:
> I think I misunderstood you, sorry. Moved it to here. Think it is a "safe" place right after LWT is published.
> 
> ```
>   if (MQTTclient.publish(LWTTopic.c_str(), LWTMessageConnect.c_str(), 1)) {
>     updateMQTTclient_connected();
>     statusLED(true);
>     mqtt_reconnect_count = 0;
>     // call all installed controller to publish autodiscover data
>     CPluginCall(CPLUGIN_GOT_CONNECTED, 0);
>     return true; // end loop if succesfull
>   }
> ```
> Should I commit the change or wait for further comments?

This one is what I meant.
You don't have to check anymore whether it was a requested reconnect or one due to connection lost. It is a reconnect, so you should call the `CPLUGIN_GOT_CONNECTED`.
And yes please add a new commit to this branch where this pull request is referring to, then it will be added to the pull request.

Christian-Me committed:
Hook for CPLUGIN_GOT_INVALID implemented & minor chnages

- hook for CPLUGIN_GOT_CONNECTED placed correctly
- hook for CPLUGIN_GOT_INVALID placed + added log
- fixed return values
- some log fixes
- cleaned up code

issues

I Think MQTTclient_should_reconnect=true does not reconnect if there is a working connection. In case of  CPLUGIN_GOT_INVALID  success a MQTTDisconnect(); function added.

Christian-Me committed:
Update platformio.ini

for dev_ESP8266_4M build

Christian-Me committed:
Merge branch 'mega' of https://github.com/Christian-Me/ESPEasy into mega

TD-er commented on platformio.ini:
```diff
@@ -12,7 +12,7 @@
 [platformio]
 ;env_default = esp32dev
-;env_default = dev_ESP8266_4096
+env_default = dev_ESP8266_4M
```
Do not commit this change

Christian-Me committed:
Revert "Update platformio.ini"

This reverts commit 3441f3519901b40cdf06fc7df0c464a10410e657.
